core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -133,13 +155,468 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+        ResidualEvaluator residuals = ResidualEvaluator.of(spec, residualFilter, isCaseSensitive());
+
+        tasks.add(
+            new BaseDeletedRowsScanTask(
+                changeOrdinal,
+                snapshot.snapshotId(),
+                dataFile.copy(shouldKeepStats()),
+                addedDeletes,
+                existingDeletes,
+                schemaString,
+                specString,
+                residuals));
+
+        // Mark this file+snapshot as processed
+        alreadyProcessedPaths.add(key);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to plan deleted rows tasks", e);
+    }
+  }
+
+  private boolean shouldKeepStats() {
+    Set<Integer> columns = columnsToKeepStats();
+    return columns != null && !columns.isEmpty();
+  }
+
+  /**
+   * Loads delete files from manifests using a cache to avoid redundant manifest parsing. This
+   * significantly improves planning performance when the same manifests are accessed across
+   * multiple scans or within a scan range.
+   *
+   * @param manifests the delete manifests to load
+   * @return list of delete files
+   */
+  private List<DeleteFile> loadDeleteFilesWithCache(List<ManifestFile> manifests) {
+    DeleteManifestCache cache = DeleteManifestCache.instance();
+    List<DeleteFile> allDeleteFiles = Lists.newArrayList();
+
+    for (ManifestFile manifest : manifests) {
+      // Try to get from cache first
+      List<DeleteFile> cachedFiles = cache.get(manifest);
+
+      if (cachedFiles != null) {
+        // Cache hit - reuse parsed delete files
+        allDeleteFiles.addAll(cachedFiles);
+      } else {
+        // Cache miss - parse manifest and populate cache
+        List<DeleteFile> manifestDeleteFiles = loadDeleteFilesFromManifest(manifest);
+        cache.put(manifest, manifestDeleteFiles);
+        allDeleteFiles.addAll(manifestDeleteFiles);
+      }
+    }
+
+    return allDeleteFiles;
+  }
+
+  /**
+   * Prunes delete manifests based on partition filter to avoid processing irrelevant manifests.
+   * This significantly improves performance when only a subset of partitions are relevant to the
+   * scan.
+   *
+   * @param manifests all delete manifests to consider
+   * @return list of manifests that might contain relevant delete files
+   */
+  private List<ManifestFile> pruneManifestsByPartition(List<ManifestFile> manifests) {
+    Expression currentFilter = filter();
+
+    // If there's no filter, return all manifests
+    if (currentFilter == null || currentFilter.equals(Expressions.alwaysTrue())) {
+      return manifests;
+    }
+
+    List<ManifestFile> prunedManifests = Lists.newArrayList();
+
+    for (ManifestFile manifest : manifests) {
+      PartitionSpec spec = table().specs().get(manifest.partitionSpecId());
+      if (spec == null || spec.isUnpartitioned()) {
+        // Include unpartitioned manifests
+        prunedManifests.add(manifest);
+        continue;
```

Comment:
nit: this control-flow feels awkward and should probably be and `else if`.

Reply:
I convered the flow to else if model.